### PR TITLE
Ship Comparison Revamp

### LINKED ISSUE #252: Rework ShipComparison page
- [x] - show unlimited ships per page (just add stupidly high number)
- [x] - hide rearrange icon and remove blocked space
- [x] - remove unnecessarily blocked space in column headers (especially the right padding)
- [ ] - Remove supership category, map to tech tree
- [x] - put all/none selectors at the front, alternatively try to align them using a table
- [x] - add tooltip to nation flags
- [ ] - use 50 rows per page by default
- [x] - remember user settings for comparison (firing range, hidden columns, rows per page, stock/top button, hide ships without section button)
- [x] - save pinned ships in URL, limit to 100, also limit in ShipStats

### PR DESCRIPTION
**Pull request by iTTou**
Implement changes listed on #252

Remove supership category and map to tech tree has to be done at data converter level. It will automatically update on the ship comparison.
It has been decided to entirely remove grid pagination.

closes #252

### CODE REVIEW

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor [hunk 1]
```diff
@@ -328,16 +448,28 @@
 
     private const string CategoriesUrlParameter = "ca";
 
-    private const string UpgradedModulesUrlParameter = "um";
+    private const string UpgradedModulesUrlParameter = "u";
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor]: this breaks backwards compatibility with older links

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor [hunk 2]
```diff
@@ -328,16 +448,28 @@
...
+
+    private const string PinnedShipsOnlyUrlParameter = "po";
+
+    private const string RangeUrlParameter = "r";
+
+    private const string HiddenColumnsUrlParameter = "hc";
+
     private Dictionary<string, bool> hiddenColumns = new();
 
+    private string hiddeColumnsUrlString = default!;
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor:280]: is this really necessary?

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor [hunk 3]
```diff
@@ -74,120 +75,233 @@
         </MudPaper>
     }
     <MudPaper Outlined Elevation="0" Class="pl-2 pb-2 mt-3">
-        <MudText GutterBottom="true" Style="font-size: medium; margin-bottom: -1px;" Class="mt-1">
-            @Localizer.SimpleAppLocalization(nameof(Translation.ShipSelectionWindow_TierFilter))
-        </MudText>
-        <MudButtonGroup OverrideStyles="false" Class="gap-2" Style="flex-wrap: wrap" Size="Size.Small">
-            @for (int i = TiersRange.First(); i <= TiersRange.Last(); i++)
+        <MudBreakpointProvider @ref="breakpointProvider">
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor]: use MudHidden instead of this

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor [hunk 4]
```diff
@@ -328,16 +448,28 @@
...
+
+    private const string HiddenColumnsUrlParameter = "hc";
+
     private Dictionary<string, bool> hiddenColumns = new();
 
+    private string hiddeColumnsUrlString = default!;
+
     private BuildConfigurationDialogHelper buildConfigurationHelper = default!;
 
+    private MudBreakpointProvider breakpointProvider = default!;
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor]: see comment about using MudHidden, then this becomes unnecessary

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor [hunk 5]
```diff
@@ -782,7 +996,28 @@
 
     private bool IsColumnHidden(string section, string columnTitle)
     {
-        return !ViewModel!.SelectedDataSection.AsString().Equals(section) || hiddenColumns.TryGetValue($"{section}_{columnTitle}", out bool hidden) && hidden;
+        return !ViewModel!.SelectedDataSection.AsString().Equals(section) || (hiddenColumns.TryGetValue($"{section}_{columnTitle}", out bool hidden) && hidden);
+    }
+
+    private bool IsSmallScreen()
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor]: Should be solved through MudHidden and not like this

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparisonViewModel.cs [hunk 6]
```diff
@@ -54,6 +54,7 @@ public ShipComparisonViewModel(ILocalizer localizer, AppSettings appSettings)
     {
         this.localizer = localizer;
         this.appSettings = appSettings;
+        LoadSettings();
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparisonViewModel.cs]: use a different name. This implies loading the settings from the local storage

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparisonViewModel.cs [hunk 7]
```diff
@@ -106,6 +107,13 @@ private static bool ContainsShipIndex(string shipIndex, IEnumerable<KeyValuePair
         return list.Select(x => x.Value.ShipIndex).Contains(shipIndex);
     }
 
+    private void LoadSettings()
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipComparison/ShipComparisonViewModel.cs]: just do this directly in the constructor

File: WoWsShipBuilder.Common/Features/ShipStats/Components/ShipStatsPanel.razor [hunk 8]
```diff
@@ -25,362 +25,388 @@
...
+                                                else
+                                                {
+                                                    foreach (var data in torpedo.DataElements)
+                                                    {
+                                                        @DataElementFragment((data, Localizer))
+                                                    }
+                                                }
+                                                @if (torpedo.CanHitClasses is not null)
+                                                {
+                                                    <div class="d-flex justify-space-between">
```
Comment (floribe2000) [inline WoWsShipBuilder.Common/Features/ShipStats/Components/ShipStatsPanel.razor]: None of these changes belong in this branch. This is unrelated to the ship comparison

**Review by floribe2000 — CHANGES_REQUESTED** [review]
There are changes completely unrelated to the ship comparison which do not belong in this PR, specifically the changes to the ShipStatsPanel. This is a separate component.
Also, either provide summaries for larger commits or describe your changes in the PR.

File: WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor (hunk 2 above)
Comment (iTTou) [reply WoWsShipBuilder.Common/Features/ShipComparison/ShipComparison.razor:280]: since generating that string requires iterating over the hidden columns dictionary, that parameter was made to calculate it only ones instead of multiple times

**Review by floribe2000 — CHANGES_REQUESTED** [review]
Most issues look good now, however, there remains one issue: Between about 1290 and 1310px window width, the button arrangment in the ship comparison selector gets messed up. Please check if you can resolve this.

**Review by floribe2000 — APPROVED** [review]
This is now ready to be merged
